src/plugin/session-tracker.ts

```diff
@@ -0,0 +1,182 @@
+/********************************************************************************
+ * Copyright (C) 2024 EclipseSource.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { DebugProtocol } from '@vscode/debugprotocol';
+import * as vscode from 'vscode';
+import { isDebugEvent, isDebugRequest, isDebugResponse } from '../common/debug-requests';
+import { WrittenMemory } from '../common/memory-range';
+
+export interface SessionInfo {
+    raw: vscode.DebugSession;
+    debugCapabilities?: DebugProtocol.Capabilities;
+    clientCapabilities?: DebugProtocol.InitializeRequestArguments;
+    active?: boolean;
+    stopped?: boolean;
```

Comment:
This one is fairly complex. Consider the code [here](https://github.com/mcu-debug/debug-tracker-vscode/blob/7a9a06a737d2537dd1be33399af174ba5b247311/src/debug-tracker.ts#L96) and my issue [here](https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/issues/325). Do we need to track this as state, or, given that we only set it true when we get a corresponding `stopped` event, could we treat that as punctual information rather than persistent state?

Reply:
Very good question! If the session would hold it's state itself (like it does in Theia) we wouldn't have the need to externally track that information and we currently use it whenever we want to update the memory views about changes in the session.

You are arguing that it should be a completely separate event that is sent to the memory views without remembering that we were stopped, right? What is the advantage of that versus remembering the state and sending it from there? I see that in your issue you mentioned that sometimes the event might not (yet) be fired so it could happen that we send "out-dated" information. However, since we always send everything we need to know about the session, that would be corrected as soon as the event is actually fired, or am I missing something?

I'm just a bit hesitant about that because the delay might be true for some sessions but it is not true for all session types and all other Debug Views also rely on the event-based mechanism as far as I can see. I'm not too strongly opposed either, I am just not sure I see the harm for now.

Reply:
I would suggest to leave it as it is here for now. I do however see @colin-grant-work 's concerns when it comes to potential multi-core debug scenarios which we need to discuss in general in the Eclipse CDT Cloud working group. There could be a need for a tracker per view instance in future. I haven't looked at all details yet, but 133 seems to make a start on that by view contexts.

As a side note: Managing such state may in future also require to look out for requests like `Continue` or `Next`. For some currently non-public extensions on our side (Core Register window), only looking out for `Continued` or `Stopped` events turned out to be problematic. The events may be fired before or after a response. And a debugger may need to back off ASAP if there is restricted access while running to avoid errors.
Also, `terminated` and `exited` events may need to be honored to stop even trying any update after exiting a session.
But let's address separately. This PR here is already quite large.